Implement ICollection<T> on several LINQ iterators

stephentoub opened the pull request:
This adds `ICollection<T>` implementations to several LINQ iterator types.  This allows operations like `list.AddRange(enumerable)` to be significantly more efficient, as AddRange queries for `ICollection<T>` and then uses its `Count`/`CopyTo` methods to directly populate the backing array.

I think the first commit here is pure goodness.  It handles Enumerable.Empty, Enumerable.Repeat, Enumerable.Range, and subsetting operations of IList, e.g. Take/Skip.

The second commit is more valuable but also may need a little more scrutiny.  It adds support for Enumerable.Select over arrays, lists, and ranges.  This makes operations like `list.AddRange(array.Select(i => ...))` much more efficient.  However, because `ICollection<T>` provides `Contains`, invoking `ICollection<T>.Contains` will also invoke the selector function for every element examined.  That's no different from if you used `Enumerable.Contains`, but in theory someone might be doing something like `if (source is ICollection<T> c) { ... issue multiple c.Contains calls }`, in which case they might silently start running the selector multiple times where they weren't previously.

eiriktsarpalis commented on src/libraries/System.Linq/tests/EmptyEnumerable.cs:
```diff
@@ -39,5 +40,27 @@ public void EmptyEnumerableIsIndeedEmpty()
+        [Fact]
+        public void ICollectionImplementationIsValid()
+        {
+            ICollection<int> c = Enumerable.Empty<int>() as ICollection<int>;
+            Assert.NotNull(c);
```
Why not use `Assert.IsAssignableFrom<ICollection<int>>(...)` here?

stephentoub replied:
Changed

eiriktsarpalis commented on src/libraries/System.Linq/src/System/Linq/Select.SpeedOpt.cs:
```diff
@@ -165,9 +165,35 @@ public TResult TryGetLast(out bool found)
                 return _selector(_source[_source.Length - 1]);
             }
+
+            bool ICollection<TResult>.Contains(TResult item)
```
I can see why you think this could be controversial. I tend to think of "collections" as structures whose elements are all materialized in-memory, and any algorithmic complexity in their operations pertains to accessing said memory. In this case however both `Contains` and `CopyTo` methods entail arbitrary work (and side-effects!), which immediately makes me classify this as an "enumerable" and not a "collection". I suspect this can lead to very difficult to diagnose regressions and it's probably not worth the perf improvements.

stephentoub replied:
Ok, I've removed the second commit for Select

eiriktsarpalis reviewed (approved):
Thanks

stephentoub committed:
Implement IList<T> on some LINQ iterators

`ICollection<T>` provides both a Count and a CopyTo, and `IList<T>` an indexer, all of which can make various consumption mechanisms more efficient. We only implement the interfaces when the underlying collection has a fixed size and all of the interface implementations are side-effect free (in particular, while appealing to do so, we don't implement them on various Select iterators).

Some of the serialization tests need to be fixed as a result. The state of Queue's array is a bit different based on how its initialized, and such private details show up in BinaryFormatter output.  Rather than special-casing the output per framework and core, I've just changed the test itself to ensure Queue can't see the size of the input collection.
